core/src/main/java/org/infinispan/interceptors/impl/QueueAsyncInvocationStage.java

```diff
@@ -126,10 +128,10 @@ private void invokeQueuedHandlers(Object rv, Throwable throwable) {
             rv = null;
             throwable = t;
          }
-         if (rv instanceof SimpleAsyncInvocationStage) {
-            SimpleAsyncInvocationStage currentStage = (SimpleAsyncInvocationStage) rv;
+         if (rv instanceof InvocationStage) {
+            InvocationStage currentStage = (InvocationStage) rv;
             if (!currentStage.isDone()) {
-               currentStage.future.whenComplete(this);
+               currentStage.toCompletableFuture().whenComplete(this);
```

Comment:
can I remove the `while(true)` then? 
I don't see any use if we are going to use the `addCallback()` every time.

Reply:
I meant always if the return value is an `InvocationStage`.

The whole `Object`/`maybeStage` thing was just a way to avoid the allocation of a `SyncInvocationStage` in every callback, so I still expect that most callbacks to not return an `InvocationStage` instance, even after the recent non-blocking work.

Reply:
ya makes sense. 
offtopic: I'm wondering if we should change the `Visitor` interface to return an `InvocationStage` instead of `Object`. 
I mean, how likely we can have an execution without having to wrap the result value in some `InvocationStage`? At least, it would avoid the `instanceOf` checks...

Reply:
It's *very* often in replicated caches without any stores, because they don't have any I/O and they can process millions of requests per second.

I have an old idea to improve this but no time to try it yet: Move the command and return value inside the context and make `InvocationContext` extend `InvocationStage<InvocationContext>`, dropping the ctx and command params in the callbacks. One of these days...